File: 146-lru-cache/lru-cache.cpp
class Node{
    public:
    int val;
    int key;

    Node*prev;
    Node*next;
public:
    Node(int key,int val)
    {
        this->key=key;
        this->val=val;
        prev=nullptr;
        next=nullptr;
    }

};
class LRUCache {
    // a doubly linkedlist and 
    // and a unordered map
    int cap;
    unordered_map<int, Node*>cache;
    Node* left;
    Node*right;  

    void insert(Node* node)
    {
        Node*prv=right->prev;
        prv->next=node;
        right->prev=node;
        node->prev=prv;
        node->next=right;

    }

    void remove(Node*node)
    {
        Node*prv=node->prev;
        Node* nxt=node->next;
        prv->next=nxt;
        nxt->prev=prv;
    }

public:

    LRUCache(int capacity) {
        cap=capacity;
        left=new Node(0,0);
        right=new Node(0,0);

        left->next=right;
        right->prev=left;
        
    }
    
    int get(int key) {
        if(cache.find(key)!=cache.end())
        {
           Node* node=cache[key];
           remove(node);
           insert(node);


            return node->val;

        }

        return -1;
    }
    
    void put(int key, int value) {
        if(cache.find(key)!= cache.end())
        {
            remove(cache[key]);

        }

        Node* node=new Node(key,value);
        cache[key]=node;
        insert(node);

        if(cache.size()>cap)
        {
            Node*lru=left->next;
            remove(lru);
            cache.erase(lru->key);

        }


        
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 146-lru-cache/lru-cache.py
class Node:
    def __init__(self,key, value):
        self.key=key
        self.value=value
        self.next=None
        self.prev=None
class LRUCache:

    def __init__(self, capacity: int):
        self.cache={}
        self.capacity=capacity
        self.left=Node(0,0)
        self.right=Node(0,0)
        self.left.next=self.right
        self.right.prev=self.left
    def _insert(self, node):
        prev=self.right.prev
        prev.next=node
        node.next=self.right
        node.prev=prev
        self.right.prev=node
    def _remove(self,node):
        prev=node.prev
        next=node.next

        prev.next=next
        next.prev=prev
        

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node=self.cache[key]

        self._remove(node)
        self._insert(node)
        return node.value
        

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node=self.cache[key]
            node.value=value
            self._remove(node)
            self._insert(node)

        else:
            self.cache[key]=Node(key,value)
        
            self._insert(self.cache[key])
            if len(self.cache)>self.capacity:
                node=self.left.next
                self._remove(node)
                del self.cache[node.key]
            
            
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)